src/instruction.rs

```diff
@@ -0,0 +1,185 @@
+//! Program instructions
+
+use solana_program::clock::UnixTimestamp;
+
+/// Instructions supported by the staking program
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum StakeInstruction {
+    /// Creates Stake config account which controls staking parameters
+    ///
+    /// 0. `[writable]` Stake config account
+    /// 1. `[]` Slash authority, may be a PDA for a governance program
+    /// 2. `[]` Config authority, may be a PDA for a governance program
+    /// 3. `[]` Stake Token Mint
+    /// 4. `[]` Stake Token Vault, to hold all staked tokens.
+    ///   Must be empty and owned by PDA with seeds `['token-owner', stake_config]`
+    InitializeConfig {
+        cooldown_time: UnixTimestamp,
+        max_deactivation_basis_points: u16,
+    },
+
+    /// Initializes stake account data for a validator.
+    ///
+    /// NOTE: Anybody can create the stake account for a validator. The authority
+    /// defaults to the validator withdraw authority.
+    ///
+    /// 0. `[]` Stake config account
+    /// 1. `[writable]` Validator stake account
+    ///     * PDA seeds: ['stake', validator, config_account]
+    /// 2. `[]` Validator vote account
+    /// 3. `[]` System program
+    InitializeStake,
+
+    /// Stakes tokens with the given config.
+    ///
+    /// Limited to the current amount of SOL staked to the validator.
```

Comment:
I think I'm a little confused on this. It looks like this instruction enables staking of some mint to the validator, but it's limited by amount of _SOL_ somehow? Can you maybe explain to me a bit more about this instruction?

Reply:
For sure! The concept is that any validator cannot have more PAL staked than their amount of SOL staked. So for example, if I have 1,000 SOL (10^12 lamports) delegated to me on the network, I cannot stake more than a total of 1,000 PAL (10^12 nano-PAL).

Reply:
Ok got it, thanks! Pegged stablecoin PAL. 😁